### PR TITLE
add define interdependent environment variables page

### PR DESCRIPTION
**Pull request by Cweiping**
Signed-off-by: Weiping Cai <weiping.cai@daocloud.io>

This PR is to fix: [#21605](https://github.com/kubernetes/website/issues/21605)
Clarify the explanation when environment variables refer to each other.And give the use case.

### CODE REVIEW

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    ```
+    SERVICE_ADDRESS
+    https://172.17.0.1:80
+    UNCHANGE_REFERENCE
+    $(PROTOCOL)://172.17.0.1:80
+    ESCAPED_REFERENCE
+    $(PROTOCOL)://172.17.0.1:80
+    ```
+
+As shown above, we have defined the correct dependency reference of `SERVICE_ADDRESS`, bad dependency reference of `UNCHANGE_REFERENCE` and skip dependent references of `ESCAPED_REFERENCE`.
```
Comment (sftim) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:68]: nit: [address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you) and avoid using “we”.

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    ```shell
+    kubectl get pods dependent-envars-demo
+    ```
+    ```
+    NAME                      READY     STATUS    RESTARTS   AGE
+    dependent-envars-demo     1/1       Running   0          9s
+    ```
+
+3. Get log to the container running in your Pod:
```
Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:54]:
```suggestion
3. Check the logs for the container running in your Pod:
```

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+## {{% heading "prerequisites" %}}
+
+
+{{< include "task-tutorial-prereqs.md" >}}
+
+
+
+
```
Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:22]: If you got a chance to send in another commit, pls remove the extra blanks lines like this.

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md (hunk 1 above)
Comment (tengqm) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:68]: As a non-native speaker, I have some problems processing the above sentence.

**Comment by Cweiping** [comment]
Unified response, thank you for your suggestions. Your suggestions have been committed, and please review again.

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md (hunk 3 above)
Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: done

**Comment by Cweiping** [comment]
> I think `UNCHANGE_REFERENCE` should be `UNCHANGED_REFERENCE`.
> 
> Other than that detail, I think this would be good to merge (I didn't yet look at the preview).

please review again.this is preview [page](https://deploy-preview-21811--kubernetes-io-master-staging.netlify.app/docs/tasks/inject-data-application/define-interdependent-environment-variables/).

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md (hunk 2 above)
Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: done

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md (hunk 1 above)
Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: done

Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]:
```suggestion
In the above example, there are three environment variables referencing the `PROTOCOL` variable.
```

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    https://172.17.0.1:80
+    UNCHANGE_REFERENCE
+    $(PROTOCOL)://172.17.0.1:80
+    ESCAPED_REFERENCE
+    $(PROTOCOL)://172.17.0.1:80
+    ```
+
+As shown above, we have defined the correct dependency reference of `SERVICE_ADDRESS`, bad dependency reference of `UNCHANGE_REFERENCE` and skip dependent references of `ESCAPED_REFERENCE`.
+
+When the environment variable is defined, you can use it directly. You can use it after the definition is completed, such as `SERVICE_ADDRESS`.
```
Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:66]:
```suggestion
When an environment variable is already defined when being referenced,
the reference can be correctly resolved, such as in the `SERVICE_ADDRESS` case.
```

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+When the environment variable is undefined or only includes some variables, the undefined environment variable is treated as a normal string, such as `UNCHANGE_REFERENCE`. A bad reference does not interfere with the operation of the container.
```
Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]:
```suggestion
When an environment variable is not yet defined when being referenced, it will be
treated as a normal string, such as in the `UNCHANGE_REFERENCE` case.
```

>  A bad reference does not interfere with the operation of the container.

I still don't get the meaning of this ... pls clarify?

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+The $(VAR_NAME) syntax can be escaped with a double $$, ie: $$(VAR_NAME).Escaped references will never be expanded, regardless of whether the variable exists or not.
```
Comment (tengqm) [inline content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md:70]:
```suggestion
The `$(VAR_NAME)` syntax can be escaped with a double `$`, ie: `$$(VAR_NAME)`.
Escaped references are never expanded, regardless of whether the referenced variable
is defined or not. This can be seen from the `ESCAPED_REFERENCE` case above.
```

File: content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md (hunk 5 above)
Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: The user-defined environment variables may not be parsed incorrectly due to the order, and will be treated as ordinary strings, but this will not affect the start of pod.

Comment (tengqm) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: so what about
```
Note that incorrectly parsed environment variables in general will not block the container from starting.
```

Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: It looks very good. Maybe my expression can't attract users' attention. Thank you for your advice.
changed it:
```
Note that incorrectly parsed environment variables, in general, will not block the container from starting.
```

Comment (Cweiping) [reply content/en/docs/tasks/inject-data-application/define-interdependent-environment-variables.md]: done. Looking forward to your review.

File: content/en/examples/pods/inject/dependent-envars.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    command:
+        - sh
+        - -c
+    image: busybox
+    env:
+    - name: SERVICE_PORT
+      value: "80"
+    - name: SERVICE_IP
+      value: "172.17.0.1"
+    - name: UNCHANGE_REFERENCE
```
Comment (sftim) [inline content/en/examples/pods/inject/dependent-envars.yaml:19]:
```suggestion
    - name: UNCHANGED_REFERENCE
```

Comment (Cweiping) [reply content/en/examples/pods/inject/dependent-envars.yaml]: Done. Sorry, missing the last change.

**Comment by tengqm** [comment]
/lgtm
Defer language proof reading to other reviewers/approvers.

File: content/en/examples/pods/inject/dependent-envars.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dependent-envars-demo
+spec:
+  containers:
+  - name: dependent-envars-demo
+    args:
+    - while true; do echo -en '\n'; echo SERVICE_ADDRESS;printenv SERVICE_ADDRESS;echo UNCHANGE_REFERENCE;printenv UNCHANGE_REFERENCE;echo ESCAPED_REFERENCE;printenv ESCAPED_REFERENCE; sleep 30; done;
```
Comment (sftim) [inline content/en/examples/pods/inject/dependent-envars.yaml:9]: `UNCHANGE_REFERENCE` should be `UNCHANGED_REFERENCE`, right?
```suggestion
    - while true; do echo -en '\n'; echo SERVICE_ADDRESS; printenv SERVICE_ADDRESS; echo UNCHANGED_REFERENCE; printenv UNCHANGED_REFERENCE; echo ESCAPED_REFERENCE; printenv ESCAPED_REFERENCE; sleep 30; done;
```

You could actually use a single `printf` there, BTW:
```suggestion
    - >
      while sleep 30; do printf "SERVICE_ADDRESS %s; UNCHANGED_REFERENCE %s; ESCAPED_REFERENCE %s\n" "${SERVICE_ADDRESS}" "${UNCHANGED_REFERENCE}" "${ESCAPED_REFERENCE}"; done;
```

:warning: _untested_

**Comment by zparnold** [comment]
@Cweiping Do you have a chance to look at @sftim 's feedback?

File: content/en/examples/pods/inject/dependent-envars.yaml (hunk 8 above)
Comment (Cweiping) [reply content/en/examples/pods/inject/dependent-envars.yaml]: Done.

**Comment by Cweiping** [comment]
Done. Looking forward to review again. @zparnold @sftim

**Comment by tengqm** [comment]
All feedbacks addressed. Thanks.
/lgtm
/approve
